jwst/assign_wcs/util.py

```diff
@@ -40,25 +41,6 @@ def __init__(self, message):
         super(MSAFileError, self).__init__(message)
 
 
-class NoDataOnDetectorError(Exception):
-    """WCS solution indicates no data on detector
-
-    When WCS solutions are available, the solutions indicate that no data
-    will be present, raise this exception.
-
-    Specific example is for NIRSpec and the NRS2 detector. For various
-    configurations of the MSA, it is possible that no dispersed spectra will
-    appear on NRS2. This is not a failure of calibration, but needs to be
-    called out in order for the calling architecture to be aware of this.
-
-    """
-
-    def __init__(self, message=None):
-        if message is None:
-            message = 'WCS solution indicate that no science is in the data.'
-        super(NoDataOnDetectorError, self).__init__(message)
```

Comment:
This exception really is specific to JWST and `assign_wcs`.  Do you think it is worthwhile having it in `stpipe`?  Is there a reason it needs to be in there?

Reply:
Ah, I see it is used in `strun`.

https://github.com/spacetelescope/jwst/blob/0382b8dbd663c4c1b61f0f91a0402ce19fe3d881/scripts/strun#L28

But this is a very specific.  Would it be better to have a more general exception to return 64 to include in `strun` checking, include that in `stpipe`, and then have this very specific one subclass it for JWST use?

Reply:
Ah I didn't realize it was specific.  How about we keep the exception in jwst but make it subclass SystemExit:

```python
class NoDataOnDetectorError(SystemExit):
    def __init__(self):
        super().__init__(64)

raise NoDataOnDetectorError()
```
After running that in a script `$?` reports 64 as expected.

Reply:
I added `SystemExit` handling to stpipe in a new PR [here](https://github.com/spacetelescope/stpipe/pull/12).  I'll update this branch to add back in `NoDataOnDetectorError` and make it subclass `SystemExit`.

Reply:
Raising `SystemExit` I think is fine.  It is a little risky to `except SystemExit`, though in the case of the CLI it should fine I think.  👍

Reply:
You make a good point in the stpipe PR that we don't need to catch it anyway...

Reply:
Thinking about this a bit more, we generally only want the `strun` CLI to be raising `SystemExit`.  And this is specifically for DMS processing.  For general user say, running some NIRSpec data through a Jupyter notebook, it would be pretty annoying to get an unhandled `SystemExit` turning off their python session (killing their notebook kernel) every time there was a no data on one of the detectors.  🤔 

Thoughts @hbushouse and @nden ?

Reply:
Hmm... another fine point.  It doesn't exit a notebook kernel but you do get an ugly message:
<img width="1125" alt="image" src="https://user-images.githubusercontent.com/2895013/107238239-9d2fed80-69f5-11eb-8d81-ae465badc14e.png">

Reply:
I added an `stpipe.exceptions.StpipeExitException` and switched the base class on `NoDataOnDetectorError` to that.

Reply:
That's a perfect solution.  🥇 

Yeah, raising `SystemExit` under the default python interpreter does exit:

```
$ python
Python 3.9.1 (default, Dec 11 2020, 06:28:49) 
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> raise SystemExit
$ 
```